modules/python/python-loghandler.c

```diff
@@ -0,0 +1,274 @@
+            (PyObject *)&PySuper_Type, (PyObject *)Py_TYPE(self), self, NULL);
+  if (!super)
+    goto exit;
+
+  init_function = _py_get_attr_or_null(super, "__init__");
+  if (!init_function)
+    goto exit;
+
+  glong level_value;
+  if (trace_flag)
+    level_value = py_log_levels.trace;
+  else if (debug_flag)
+    level_value = py_log_levels.debug;
+  else
+    level_value = py_log_levels.info;
+
+  PyObject *init_ret = PyObject_CallFunction(init_function, "(l)", level_value);
+  ret = init_ret ? 0 : -1;
+  Py_DECREF(init_ret);
+
+exit:
+  Py_XDECREF(logging_module);
+  Py_XDECREF(super);
+  Py_XDECREF(init_function);
+
+  return ret;
+}
+
+PyObject *py_loghandler_emit(PyObject *self, PyObject *args)
+{
+  PyObject *formatted = NULL;
+
+  PyObject *record;
+  if (!PyArg_ParseTuple(args, "O", &record))
+    return NULL;
+
+  formatted = PyObject_CallMethod(self, "format", "(O)", record);
+  if (!formatted)
+    goto exit_error;
+
+  const char *message = PyUnicode_AsUTF8AndSize(formatted, NULL);
+  if (!message)
+    goto exit_error;
+
+  glong level = _py_fetch_log_level(record, "levelno");
+  if (level < 0)
+    goto exit_normal;
+
+  if (level >= py_log_levels.error)
+    msg_error(message);
+  else if (level >= py_log_levels.warning)
+    msg_warning(message);
+  else if (level >= py_log_levels.info)
+    msg_info(message);
+  else if (level >= py_log_levels.debug && debug_flag)
+    msg_debug(message);
+  else if (level >= py_log_levels.trace && trace_flag)
+    msg_trace(message);
+
+exit_normal:
+  Py_XDECREF(formatted);
+  Py_RETURN_NONE;
+
+exit_error:
+  Py_XDECREF(formatted);
+  return NULL;
+}
+
+static PyMethodDef py_handler_methods[] =
+{
+  {"emit", (PyCFunction)py_loghandler_emit, METH_VARARGS, "emit a log record"},
+  {NULL}
+};
+
+static PyType_Slot py_handler_slots[] =
+{
+  {Py_tp_doc,     "Those messages can be captured by internal() source."},
+  {Py_tp_new,     PyType_GenericNew},
+  {Py_tp_init, (initproc)py_loghandler_init},
+  {Py_tp_dealloc,  (destructor)py_slng_generic_dealloc},
+  {Py_tp_methods, py_handler_methods},
+  {0, 0}
+};
+
+static PyType_Spec py_handler_spec =
+{
+  .name = "InternalHandler",
+  .basicsize = sizeof(PyObject),
+  .flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_BASETYPE,
+  .slots = py_handler_slots,
+};
+
+void py_loghandler_global_init(void)
+{
+  PyObject *logging_module = NULL;
+  PyObject *bases = NULL;
+
+  logging_module = _py_do_import("logging");
+  if (!logging_module)
+    goto exit;
+
+  PyObject *handler = PyObject_GetAttrString(logging_module, "Handler");
+  if (!handler)
+    {
+      gchar buf[256];
+      msg_error(
+        "could not find logging.Handler class",
+        evt_tag_str("exception", _py_format_exception_text(buf, sizeof(buf))));
+      goto exit;
+    }
+
+  bases = PyTuple_Pack(1, handler);
+  if (!bases)
+    {
+      Py_XDECREF(handler);
+      goto exit;
+    }
+
+  PyObject *internal_handler = PyType_FromSpecWithBases(&py_handler_spec, bases);
```

Comment:
```suggestion
    *internal_handler = PyType_FromSpecWithBases(&py_handler_spec, bases);
  ```
  to decrement it's ref inside the `exit` label to avoid mem leakage